src/RootComponent.js

```diff
@@ -0,0 +1,61 @@
+import React from 'react';
+import Relay from 'react-relay';
+
+import QueryAggregator from './QueryAggregator';
+import StemContainer from './StemContainer';
+
+export default class RootComponent extends React.Component {
+  static propTypes = {
+    routerProps: React.PropTypes.object.isRequired,
+  };
+
+  static childContextTypes = {
+    queryAggregator: React.PropTypes.instanceOf(QueryAggregator).isRequired,
+  };
+
+  constructor(props, context) {
+    super(props, context);
+
+    this.queryAggregator = new QueryAggregator();
+    this.queryAggregator.updateRoute(props.routerProps);
+  }
+
+  getChildContext() {
+    return {
+      queryAggregator: this.queryAggregator,
+    };
+  }
+
+  componentWillReceiveProps(nextProps) {
+    const {routerProps} = nextProps;
+    if (routerProps.isTransitioning) {
+      return;
+    }
+
+    this.queryAggregator.updateRoute(routerProps);
+  }
+
+  renderLoading = () => {
```

Comment:
That also breaks hypothetical pure render optimizations though. Would prefer to preserve referential equality when possible.